src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+                }
+              }
+              // Append to the existing globs, and collate at the end of this iteration.
+              let entry = todo_directories
+                .entry(directory_path)
+                .or_insert_with(Vec::new);
+              entry.extend(subdir_globs);
+            }
+          }
+        }
+      }
+
+      // 1c. Push any "dormant" contexts to traverse subdirs onto the dormant stack.
+      let dependencies: Vec<PathBuf> = {
+        let mut dependencies: Vec<PathBuf> = vec![];
+        // If we ever encounter a directory *with* a remainder, we have to ensure that that
+        // directory is *not* in `globbed_directories`, which are *not* subsetted at all.
+        // NB: Because our goal is to *union* all of the includes, if a directory is in
+        // `globbed_directories`, we can skip processing it for subsetting here!
+        for subdir_path in todo_directories.keys() {
+          if !globbed_directories.contains_key(subdir_path) {
+            dependencies.push(prefix.join(subdir_path));
+          }
+        }
+        dependencies
+      };
+
+      for (subdir_name, all_path_globs) in todo_directories.into_iter() {
+        let full_name = prefix.join(&subdir_name);
+        let bazel_digest = cur_dir_directories
+          .get(&subdir_name)
+          .unwrap()
+          .get_digest()
+          .clone();
+        let digest = to_pants_digest(bazel_digest);
+        let multiple_globs = MultipleGlobs {
+          include: all_path_globs,
+          exclude: exclude.clone(),
+        };
+        unexpanded_stack.insert(
+          full_name,
+          UnexpandedSubdirectoryContext {
+            digest,
+            multiple_globs,
+          },
+        );
+      }
+
+      // 1d. Push a base "activated" context for the current directory node, then push dormant
+      // nodes for the subdirectories to glob.
+      let activated_context = PartiallyExpandedDirectoryContext {
+        files: globbed_files.into_iter().map(|(_, node)| node).collect(),
+        known_directories: globbed_directories
+          .into_iter()
+          .map(|(_, node)| node)
+          .collect(),
+        directory_promises: dependencies,
+      };
+      partially_expanded_stack.insert(prefix, activated_context);
+    }
+
+    // 2. Zip back up the recursively subsetted directory protos.
+    let mut completed_digests: IndexMap<PathBuf, Digest> = IndexMap::new();
+    while let Some((
+      prefix,
+      PartiallyExpandedDirectoryContext {
+        files,
+        known_directories,
+        directory_promises,
+      },
+    )) = partially_expanded_stack.pop()
+    {
+      let completed_nodes: Vec<remexec::DirectoryNode> = directory_promises
+        .into_iter()
+        .map(|dependency| {
+          let digest = completed_digests.get(&dependency).ok_or_else(|| {
+            format!(
+              "expected dependency {:?} to be available from completed_digests {:?} -- internal error",
```

Comment:
Maybe rather than "dependencies" these are child nodes?

Reply:
They are both! I have added a comment here to make it more clear!